src/onegov/search/mixins.py

```diff
@@ -134,6 +136,139 @@ def es_tags(self):
+        :return: tsvector string
+        """
+
+        join = " || ' ' || "
+
+        # identify search columns
+        columns = [p for p in model.es_properties if p in model.__dict__
+                   and isinstance(model.__dict__[p], InstrumentedAttribute)]
+        s = Searchable.create_tsvector_string(*columns) if columns else ''
+
+        # identify content and meta properties
+        for p in model.es_properties:
+            if p in model.__dict__ and isinstance(model.__dict__[p],
+                                                  content.dict_property):
+                s += join if s else ''
+                s += f"coalesce((({model.__dict__[p].attribute} " \
+                     f"->> '{p}')), '')"
+
+        return s
+
+    @staticmethod
+    def reindex(session, schema, model):
+        """
+        Re-indexes the table by dropping and adding the full text search
+        column.
+        """
+        Searchable.drop_fts_column(session, schema, model)
+        Searchable.add_fts_column(session, schema, model)
+
+    @staticmethod
+    def drop_fts_column(session, schema, model, col_name='fts_idx'):
+        """
+        Drops the full text search column
+
+        :param session: db session
+        :param schema: schema the full text column shall be added
+        :param model: model to drop the index from
+        :param col_name: column name of fts index
+        :return: None
+        """
+        query = f"""
+            ALTER TABLE "{schema}".{model.__tablename__} DROP COLUMN IF EXISTS
+            {col_name};
+        """
+        session.execute(query)
+        session.execute("COMMIT")
+
+    @staticmethod
+    def add_fts_column(session, schema, model, col_name='fts_idx'):
+
+        """
+        This function is used for re-indexing and as migration step moving to
+        postgresql full text search (fts), OGC-508.
+
+        It adds a separate column for the tsvector to `schema`.`table`
+        creating a multilingual gin index on the columns/data defined per
+        model.
+
+        :param session: session
+        :param schema: schema name
+        :param model: model to add the index
+        :param col_name: column name of fts index
+        :return: None
+        """
+
+        # we build the tsvector expression for all supported languages based
+        # on this format
+        # to_tsvector('simple', '<tsvector_string_of_model') ||
+        # to_tsvector('german', '<tsvector_string_of_model') ||
+        # to_tsvector('french', '<tsvector_string_of_model') ||
+        # ...
+
+        tsvector_expression = \
+            Searchable.multi_language_tsvector_expression(
+                model.psql_tsvector_string(model))
+
+        query = f"""
+            ALTER TABLE "{schema}".{model.__tablename__}
+            ADD COLUMN IF NOT EXISTS {col_name} tsvector GENERATED ALWAYS AS
+            ({tsvector_expression}) STORED;
+        """
+        session.execute(query)
+        session.execute("COMMIT")
```

Comment:
We probably need to do something more sophisticated in here to properly handle polymoprhic models.
1.  When you determine all the searchable models get their polymorphic base class and use that to deduplicate the polymorphic variants of the models and only pass in non-polymorphic or polypmorphic base classes into this function.
2. Inside this function retrieve all the polymorphic subclasses which derive from Searchable.
3. Create a CASE statement which maps each polymorphic model's ts_vector_expression to the polymorphic identity

Steps 2/3 are only necessary if the model is polymorphic, otherwise you can use your simple version.

You can probably simplify 3 a little bit by merging all the cases which produce an identical tsvector_statement. The Postgres query compiler will probably do that for you, but we probably should still try to generate the most simple expression possible.

Reply:
I don't get your idea fully. 

Note that I check the __tablename__ to ensure only re-indexing each table once...  -> see ../onegov-cloud/src/onegov/search/integration.py +L401 in psql_perform_reindex()

Reply:
Consider the following example:
```
class User(Base):
    __tablename__ = 'users'
    __mapper__args = {'polymorphic_on': 'type', 'polymorphic_identity': 'generic'}
    type = Column(String, nullable=False, default=lambda: 'generic')
    username = Column(String, nullable=False)
    email = Column(String, nullable=False)

class Person(User, Searchable):
    es_properties = {'username': {'type': 'text'}}
    es_public = True

class Participant(User, Searchable):
    es_properties = {'username': {'type': 'text'}, 'email': {'type': 'text'}}
    es_public = False
```

Now you have three polymorphic variations that each have their own requirements. The base class isn't supposed to show up in results at all, the two derived classes are searchable, but the first only allows searching for the username, while the other also allows searching for email but as a trade-off the results are only visible to logged in users.

With your naive approach of generating one ts_vector_expression per table you don't know which expression it will pick up, and the results will be wrong for all the polymorphic variants it did not pick.

You can use `sqlalchemy.inspect` to get the `Mapper` associated with the model, which will give you access to properties such as `polymorphic_identity` and `self_and_descendants` to get all the polymorphic variations of the class, starting from the current variant.

Reply:
Weird, I did exactly this otherwise re-indexing would not work!